olp-cpp-sdk-core/src/client/OlpClient.cpp

```diff
@@ -537,103 +466,49 @@ HttpResponse OlpClient::OlpClientImpl::CallApi(
-    if (CaseInsensitiveCompare(header.first, http::kUserAgentHeader)) {
-      custom_user_agent += header.second + std::string(" ");
-    } else {
-      network_request.WithHeader(header.first, header.second);
-    }
-  }
-
-  custom_user_agent += http::kOlpSdkUserAgent;
-  network_request.WithHeader(http::kUserAgentHeader, custom_user_agent);
-
-  if (!content_type.empty()) {
-    network_request.WithHeader(http::kContentTypeHeader, content_type);
-  }
-
-  auto backdown_period =
-      std::chrono::milliseconds(retry_settings.initial_backdown_period);
-
-  AddBearer(query_params.empty(), network_request);
-
-  auto response =
-      SendRequest(network_request, settings_, retry_settings, context);
-
-  NetworkStatistics accumulated_statistics = response.GetNetworkStatistics();
-
-  // Make sure that we don't wait longer than `timeout` in retry settings
-  auto accumulated_wait_time = backdown_period;
-  const auto max_wait_time =
-      std::chrono::duration_cast<std::chrono::milliseconds>(
-          std::chrono::seconds(retry_settings.timeout));
-  for (int i = 1; i <= retry_settings.max_attempts && !context.IsCancelled() &&
-                  accumulated_wait_time < max_wait_time;
-       i++) {
-    if (StatusSuccess(response.status)) {
-      return response;
-    }
-
-    if (!retry_settings.retry_condition(response)) {
-      return response;
-    }
+  struct ResponseData {
+    Condition condition;
+    HttpResponse response = ToHttpResponse(kCancelledErrorResponse);
+  };
 
-    // do the periodical sleep and check for cancellation status in between.
-    auto duration_to_sleep =
-        std::min(backdown_period, max_wait_time - accumulated_wait_time);
-    accumulated_wait_time += duration_to_sleep;
+  auto response_data = std::make_shared<ResponseData>();
 
-    while (duration_to_sleep.count() > 0 && !context.IsCancelled()) {
-      const auto sleep_ms =
-          std::min(std::chrono::milliseconds(1000), duration_to_sleep);
-      std::this_thread::sleep_for(sleep_ms);
-      duration_to_sleep -= sleep_ms;
-    }
+  auto callback = [=](HttpResponse http_response) {
+    response_data->response = std::move(http_response);
+    response_data->condition.Notify();
+  };
 
-    backdown_period = CalculateNextWaitTime(retry_settings, i);
-    response = SendRequest(network_request, settings_, retry_settings, context);
+  context.ExecuteOrCancelled(
+      [&]() -> client::CancellationToken {
+        auto token = CallApi(path, method, query_params, header_params,
+                             forms_params, post_body, content_type, callback);
+        return CancellationToken([=] {
+          token.Cancel();
+          response_data->condition.Notify();
+        });
+      },
+      [=]() { response_data->condition.Notify(); });
 
-    // In case we retry, accumulate the stats
-    accumulated_statistics += response.GetNetworkStatistics();
+  auto waiting_time = settings_.retry_settings.timeout *
```

Comment:
The `timeout` is the total timeout for all retries, which means you should not multiple it. E.g. if the user specifies timeout = 60s, then the total expected time, incl. all retries and waits, should not exceed 60s.